### PR TITLE
Online LARFT

### LINKED ISSUE #1339: [Tools] Online LARFT
## Demo
[HMDA Beta - LAR Formatting Tool](https://ffiec.beta.cfpb.gov/tools/lar-formatting)
 
[Latest Documentation](https://github.com/cfpb/hmda-frontend/blob/5083877a1ff0bb4cba9069bdc500e3a11a5cb5a7/src/documentation/markdown/2021/lar-formatting-new.md) (Updated: 2/25/22)

## Motivation
The current LAR Formatting Tool is Excel based and does not work effectively on OSX. 

## Goals
As an MVP, we want to provide a web-based tool to 
- CRUD Rows
- Export a LAR file
- Import a LAR file

## Longterm we'd like to
- Integrate the check-digit calculator
- Integrate the rate spread calculator
- Integrate single-line edit checks
- Integrate full-file edit checks
- Optimize performance for larger LAR files

## Bugs
- `Age` fields need to allow numeric entry, along with the NA/No-co applicant entries in the dropdown
- `Credit score` need to allow numeric input

### PR DESCRIPTION
**Pull request by meissadia**
Closes #1339 

## Changes

- CRUD of Transmittal Sheet and LAR rows
- Upload and edit pipe-delimited LAR file
- Enter, edit, and export LAR data as a pipe-delimited LAR file
- Easily identify which column you are editing by name
- Pasting a pipe-delimited row into the text area parses the row
- Performance optimizations to handle large quantities of records
- Buttons to Copy from/Paste to text area without having to click-to-focus, select all, paste/copy.
- Confirmation dialogs for destructive actions (delete, delete all [clear], overwrite all [upload])
- Field-specific example is provided as the `placeholder` for text fields
- Ability to select from a pre-defined list for fields with strict enumerations
- Easy date selection

## Testing

1. 15,000 row file 
  - Able to easily scroll but still hard to find a specific row with that many entries
  - Crashes sometimes
  - Editing a row is slow to update the `selectedRow`
  - Needs heavy performance optimization for larger datasets
2. 10 row file
  - Jumpy scrolling when deleting a row (where should the user focus next?)
  - provide easier navigation between sections (ts/lar/raw/parsed)?
    - https://mckervinc.github.io/react-fluid-table/#/scroll

## Screenshots
screenshot

### CODE REVIEW

**Comment by meissadia** [comment]
Excluded rows: 
Screen Shot 2022-02-24 at 3 08 05 PM

File: src/documentation/markdown/2023/lar-formatting.md [hunk 1]
```diff
@@ -1,137 +1,165 @@
...
+### File Actions
 
-3. If the message “This file originated from an Internet location and might be unsafe” appears, select the **“Enable Editing”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Enable_Editing.png)
+Import or export your LAR data.
 
-4. If the message “Macros have been disabled” appears, select the **“Enable Content”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Macros_enabled.png)
+- **Upload**: Work with an existing pipe-delimited LAR file.
+- **Download**: Create a [LAR file](#lar-file) based on the currently saved LAR data.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2023/lar-formatting.md:10]: The`[LAR file](#lar-file)` isn't link to anything in the UI.

File: src/documentation/markdown/2022/lar-formatting.md [hunk 2]
```diff
@@ -1,137 +1,165 @@
...
+### File Actions
 
-3. If the message “This file originated from an Internet location and might be unsafe” appears, select the **“Enable Editing”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Enable_Editing.png)
+Import or export your LAR data.
 
-4. If the message “Macros have been disabled” appears, select the **“Enable Content”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Macros_enabled.png)
+- **Upload**: Work with an existing pipe-delimited LAR file.
+- **Download**: Create a [LAR file](#lar-file) based on the currently saved LAR data.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2022/lar-formatting.md:10]: The`[LAR file](#lar-file)` isn't link to anything in the UI.

File: src/documentation/markdown/2021/lar-formatting.md [hunk 3]
```diff
@@ -1,137 +1,165 @@
...
+### File Actions
 
-3. If the message “This file originated from an Internet location and might be unsafe” appears, select the **“Enable Editing”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Enable_Editing.png)
+Import or export your LAR data.
 
-4. If the message “Macros have been disabled” appears, select the **“Enable Content”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Macros_enabled.png)
+- **Upload**: Work with an existing pipe-delimited LAR file.
+- **Download**: Create a [LAR file](#lar-file) based on the currently saved LAR data.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2021/lar-formatting.md:10]: The`[LAR file](#lar-file)` isn't link to anything in the UI.

File: src/documentation/markdown/2020/lar-formatting.md [hunk 4]
```diff
@@ -1,137 +1,165 @@
...
+### File Actions
 
-3. If the message “This file originated from an Internet location and might be unsafe” appears, select the **“Enable Editing”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Enable_Editing.png)
+Import or export your LAR data.
 
-4. If the message “Macros have been disabled” appears, select the **“Enable Content”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Macros_enabled.png)
+- **Upload**: Work with an existing pipe-delimited LAR file.
+- **Download**: Create a [LAR file](#lar-file) based on the currently saved LAR data.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2020/lar-formatting.md:10]: The`[LAR file](#lar-file)` isn't link to anything in the UI.

File: src/documentation/markdown/2019/lar-formatting.md [hunk 5]
```diff
@@ -1,137 +1,165 @@
...
+### File Actions
 
-3. If the message “This file originated from an Internet location and might be unsafe” appears, select the **“Enable Editing”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Enable_Editing.png)
+Import or export your LAR data.
 
-4. If the message “Macros have been disabled” appears, select the **“Enable Content”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Macros_enabled.png)
+- **Upload**: Work with an existing pipe-delimited LAR file.
+- **Download**: Create a [LAR file](#lar-file) based on the currently saved LAR data.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2019/lar-formatting.md:10]: The`[LAR file](#lar-file)` isn't link to anything in the UI.

File: src/documentation/markdown/2019/lar-formatting.md [hunk 6]
```diff
@@ -1,137 +1,165 @@
...
+  - [`Parsed Values`](#parsed-values) provides examples of field inputs, field and input descriptions, and easy selectors for enumerated fields.
+  - [`Pipe-Delimited Values`](#pipe-delimited-values) provides the row's raw text, as it will appear in your downloaded LAR file.
+    - Clicking in a field will highlight that value in the [`Parsed Values`](#parsed-values) section for quick editing.
 
-7.	Select **OK** to close the Security dialog box.
+### Save changes
+After making your changes, use the [`Download File`](#file-actions) button to download your LAR file.
 
-8.	Select **OK** to close the Options dialog box.
+The data is now formatted into a pipe-delimited text file ([as shown below](#lar-file-output)) and ready to be submitted to the HMDA Platform.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2019/lar-formatting.md:147]: `[as shown below](#lar-file-output)` doesn't link to anything.

File: src/documentation/markdown/2018/lar-formatting.md [hunk 7]
```diff
@@ -1,137 +1,165 @@
...
+### File Actions
 
-3. If the message “This file originated from an Internet location and might be unsafe” appears, select the **“Enable Editing”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Enable_Editing.png)
+Import or export your LAR data.
 
-4. If the message “Macros have been disabled” appears, select the **“Enable Content”** button.
-![](https://raw.githubusercontent.com/cfpb/hmda-frontend/master/src/documentation/markdown/images/larft/Macros_enabled.png)
+- **Upload**: Work with an existing pipe-delimited LAR file.
+- **Download**: Create a [LAR file](#lar-file) based on the currently saved LAR data.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2018/lar-formatting.md:10]: The `[LAR file](#lar-file)` does not link to anything.

File: src/documentation/markdown/2018/lar-formatting.md [hunk 8]
```diff
@@ -1,137 +1,165 @@
...
+  - [`Parsed Values`](#parsed-values) provides examples of field inputs, field and input descriptions, and easy selectors for enumerated fields.
+  - [`Pipe-Delimited Values`](#pipe-delimited-values) provides the row's raw text, as it will appear in your downloaded LAR file.
+    - Clicking in a field will highlight that value in the [`Parsed Values`](#parsed-values) section for quick editing.
 
-7.	Select **OK** to close the Security dialog box.
+### Save changes
+After making your changes, use the [`Download File`](#file-actions) button to download your LAR file.
 
-8.	Select **OK** to close the Options dialog box.
+The data is now formatted into a pipe-delimited text file ([as shown below](#lar-file-output)) and ready to be submitted to the HMDA Platform.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2018/lar-formatting.md:147]: `[as shown below](#lar-file-output)` doesn't link to anything.

File: src/documentation/markdown/2020/lar-formatting.md [hunk 9]
```diff
@@ -1,137 +1,165 @@
...
+  - [`Parsed Values`](#parsed-values) provides examples of field inputs, field and input descriptions, and easy selectors for enumerated fields.
+  - [`Pipe-Delimited Values`](#pipe-delimited-values) provides the row's raw text, as it will appear in your downloaded LAR file.
+    - Clicking in a field will highlight that value in the [`Parsed Values`](#parsed-values) section for quick editing.
 
-7.	Select **OK** to close the Security dialog box.
+### Save changes
+After making your changes, use the [`Download File`](#file-actions) button to download your LAR file.
 
-8.	Select **OK** to close the Options dialog box.
+The data is now formatted into a pipe-delimited text file ([as shown below](#lar-file-output)) and ready to be submitted to the HMDA Platform.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2020/lar-formatting.md:147]: `[as shown below](#lar-file-output)` doesn't link to anything.

File: src/documentation/markdown/2021/lar-formatting.md [hunk 10]
```diff
@@ -1,137 +1,165 @@
...
 
-8.	Select **OK** to close the Options dialog box.
+The data is now formatted into a pipe-delimited text file ([as shown below](#lar-file-output)) and ready to be submitted to the HMDA Platform.
 
-9.	Return to step 4 in the instructions for the HMDA LAR Formatting Tool.
- 
+## File Output
 
-### Transferring existing data from another workbook or Google Sheets™ web-based spreadsheet program
+The LARFT produces pipe-delimited, UTF-8 encoded LAR files. Files contain at least 1 TS and 1 LAR row.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2021/lar-formatting.md:151]: `[as shown below](#lar-file-output)` doesn't link to anything.

File: src/documentation/markdown/2022/lar-formatting.md [hunk 11]
```diff
@@ -1,137 +1,165 @@
...
+  - [`Parsed Values`](#parsed-values) provides examples of field inputs, field and input descriptions, and easy selectors for enumerated fields.
+  - [`Pipe-Delimited Values`](#pipe-delimited-values) provides the row's raw text, as it will appear in your downloaded LAR file.
+    - Clicking in a field will highlight that value in the [`Parsed Values`](#parsed-values) section for quick editing.
 
-7.	Select **OK** to close the Security dialog box.
+### Save changes
+After making your changes, use the [`Download File`](#file-actions) button to download your LAR file.
 
-8.	Select **OK** to close the Options dialog box.
+The data is now formatted into a pipe-delimited text file ([as shown below](#lar-file-output)) and ready to be submitted to the HMDA Platform.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2022/lar-formatting.md:147]: `[as shown below](#lar-file-output)` doesn't link to anything.

File: src/documentation/markdown/2023/lar-formatting.md [hunk 12]
```diff
@@ -1,137 +1,165 @@
...
+  - [`Parsed Values`](#parsed-values) provides examples of field inputs, field and input descriptions, and easy selectors for enumerated fields.
+  - [`Pipe-Delimited Values`](#pipe-delimited-values) provides the row's raw text, as it will appear in your downloaded LAR file.
+    - Clicking in a field will highlight that value in the [`Parsed Values`](#parsed-values) section for quick editing.
 
-7.	Select **OK** to close the Security dialog box.
+### Save changes
+After making your changes, use the [`Download File`](#file-actions) button to download your LAR file.
 
-8.	Select **OK** to close the Options dialog box.
+The data is now formatted into a pipe-delimited text file ([as shown below](#lar-file-output)) and ready to be submitted to the HMDA Platform.
```
Comment (Michaeldremy) [inline src/documentation/markdown/2023/lar-formatting.md:147]: `[as shown below](#lar-file-output)` doesn't link to anything.

**Review by Michaeldremy — CHANGES_REQUESTED** [review]
It would be helpful to include a readme that discusses the overall flow of how the data is handled. Reason for this would be to quickly identify where an issue might end up being. There are a lot of moving pieces/components and to have a visual table of contents discussing the data flow would be extremely helpful.

Additionally I think it would be good to add some general documentation on a lot of the components, to briefly explain what they do. Use case is when hovering over a component it gives a [JSDoc](https://jsdoc.app/) tooltip explanation.

File: src/documentation/markdown/2023/lar-formatting.md (hunk 1 above)
Comment (meissadia) [reply src/documentation/markdown/2023/lar-formatting.md:10]: Updated

File: src/documentation/markdown/2020/lar-formatting.md (hunk 9 above)
Comment (meissadia) [reply src/documentation/markdown/2020/lar-formatting.md:147]: updated

File: src/documentation/markdown/2021/lar-formatting.md (hunk 10 above)
Comment (meissadia) [reply src/documentation/markdown/2021/lar-formatting.md:151]: updated

File: src/documentation/markdown/2023/lar-formatting.md (hunk 12 above)
Comment (meissadia) [reply src/documentation/markdown/2023/lar-formatting.md:147]: updated

File: src/documentation/markdown/2022/lar-formatting.md (hunk 11 above)
Comment (meissadia) [reply src/documentation/markdown/2022/lar-formatting.md:147]: updated

File: src/documentation/markdown/2018/lar-formatting.md (hunk 8 above)
Comment (meissadia) [reply src/documentation/markdown/2018/lar-formatting.md:147]: updated

File: src/documentation/markdown/2022/lar-formatting.md (hunk 2 above)
Comment (meissadia) [reply src/documentation/markdown/2022/lar-formatting.md:10]: updated

File: src/documentation/markdown/2020/lar-formatting.md (hunk 4 above)
Comment (meissadia) [reply src/documentation/markdown/2020/lar-formatting.md:10]: updated

File: src/documentation/markdown/2019/lar-formatting.md (hunk 6 above)
Comment (meissadia) [reply src/documentation/markdown/2019/lar-formatting.md:147]: updated

File: src/documentation/markdown/2018/lar-formatting.md (hunk 7 above)
Comment (meissadia) [reply src/documentation/markdown/2018/lar-formatting.md:10]: updated

File: src/tools/larft/components/saved-rows/buildColumns.jsx [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+// Build element ID from field name
+const formatFieldID = f => {
+  const replaceCharacters = [':', ' ', "'", ',', '(', ')']
+  let adjusted = f.fieldName.toLowerCase()
+  replaceCharacters.forEach(replaceable => {
+    adjusted = adjusted.replaceAll(replaceable, '-')
+  })
+  return 'header-' + adjusted
+}
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/tools/larft/components/saved-rows/buildColumns.jsx:132]: Implemented similar.  Regex needed the global flag `/g` and we need to `replaceAll` instances. 

```
const formatFieldID = f => {
  const REGEX_UNWANTED = /[:\s',()]/g
  let adjusted = f.fieldName.toLowerCase().replaceAll(REGEX_UNWANTED, '-')
  return 'header-' + adjusted
}
```

File: src/tools/larft/components/saved-rows/buildColumns.jsx [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+// ----------- Utils -----------
+
+const filterFields = ({ matchedColumns, schema, columnFilter }) =>
+  schema
+    .filter(x => {
+      if (!matchedColumns.length) return true
+      if (matchedColumns.includes(x.fieldName)) return true
+    })
+    .filter(x => matchColumnFilter(x, columnFilter))
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/tools/larft/components/saved-rows/buildColumns.jsx:124]: Added

**Review by Michaeldremy — APPROVED** [review]
Code looks good to me. 

Still want to throw out my previous overarching comment and post here.

> It would be helpful to include a readme that discusses the overall flow of how the data is handled. Reason for this would be to quickly identify where an issue might end up being. There are a lot of moving pieces/components and to have a visual table of contents discussing the data flow would be extremely helpful.

> Additionally I think it would be good to add some general documentation on a lot of the components, to briefly explain what they do. Use case is when hovering over a component it gives a [JSDoc](https://jsdoc.app/) tooltip explanation.

**Comment by meissadia** [comment]
@Michaeldremy First draft of some dev docs for OLARFT.  

Would love some feedback on what is useful, what should be changed/removed, what could be added.

https://github.com/cfpb/hmda-frontend/blob/1339-olarft/src/tools/larft/LARFT%20Dev%20Docs.md

**Comment by meissadia** [comment]
[Dev docs for OLARFT](https://github.com/cfpb/hmda-frontend/blob/1339-olarft/src/tools/larft/LARFT%20Dev%20Docs.md)

**Pull request merged**